libs/x-content/src/main/java/org/elasticsearch/xcontent/ObjectParser.java

```diff
@@ -289,36 +286,32 @@ public Value parse(XContentParser parser, Value value, Context context) throws I
             }
         }
 
+        FieldParser fieldParser;
+        String currentFieldName;
+        XContentLocation currentPosition;
         final Map<String, FieldParser> parsers = fieldParserMap.getOrDefault(parser.getRestApiVersion(), Collections.emptyMap());
-        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
-            if (token == XContentParser.Token.FIELD_NAME) {
-                currentFieldName = parser.currentName();
-                currentPosition = parser.getTokenLocation();
-                fieldParser = parsers.get(currentFieldName);
+        while ((currentFieldName = parser.nextFieldName()) != null) {
+            currentPosition = parser.getTokenLocation();
+            fieldParser = parsers.get(currentFieldName);
+            token = parser.nextToken();
+            if (fieldParser == null) {
+                unknownFieldParser.acceptUnknownField(this, currentFieldName, currentPosition, parser, value, context);
             } else {
-                if (currentFieldName == null) {
-                    throwNoFieldFound(parser);
```

Comment:
I'm wondering about parsing '{}' START_OBJECT END_OBJECT, would we throw no field found before and now successfully parse?

Reply:
I don't think so because because we'd never enter the loop because the `nextToken` call would return `END_OBJECT` right away right?

Reply:
Yeah, you are right. I wrote a little test to convince myself this morning, no behaviour change:

```
    public void testEmpty() throws IOException {
        class TestStruct {
        }

        ObjectParser<TestStruct, Void> objectParser = new ObjectParser<>("the_parser");
        TestStruct s = new TestStruct();

        {
            XContentParser parser = createParser(JsonXContent.jsonXContent, "{}");
            objectParser.parse(parser, s,null);
        }
    }
```